server/src/main/java/org/apache/iotdb/db/engine/compaction/io/CompactionTsFileReader.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.iotdb.tsfile.read.TsFileDeviceIterator;
+import org.apache.iotdb.tsfile.read.TsFileSequenceReader;
+import org.apache.iotdb.tsfile.read.common.Chunk;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class extends the TsFileSequenceReader class to read and manage TsFile with a focus on
+ * compaction-related operations. This includes functions for tracking and recording the amount of
+ * data read and distinguishing between aligned and not aligned series during compaction.
+ */
+public class CompactionTsFileReader extends TsFileSequenceReader {
+  /** Tracks the total amount of data (in bytes) that has been read. */
+  private AtomicLong readDataSize = new AtomicLong(0L);
+
+  /** The type of compaction running. */
+  CompactionType compactionType;
+
+  /** A flag that indicates if an aligned series is being read. */
+  private volatile boolean readingAlignedSeries = false;
+
+  /**
+   * Constructs a new instance of CompactionTsFileReader.
+   *
+   * @param file The file to be read.
+   * @param compactionType The type of compaction running.
+   * @throws IOException If an error occurs during file operations.
+   */
+  public CompactionTsFileReader(String file, CompactionType compactionType) throws IOException {
+    super(file);
+    this.compactionType = compactionType;
+  }
+
+  @Override
+  protected ByteBuffer readData(long position, int totalSize) throws IOException {
+    ByteBuffer buffer = super.readData(position, totalSize);
+    readDataSize.addAndGet(totalSize);
+    return buffer;
+  }
+
+  /** Marks the start of reading an aligned series. */
+  public void markStartOfAlignedSeries() {
+    readingAlignedSeries = true;
+  }
+
+  /** Marks the end of reading an aligned series. */
+  public void markEndOfAlignedSeries() {
+    readingAlignedSeries = false;
+  }
+
+  @Override
+  public Chunk readMemChunk(ChunkMetadata metaData) throws IOException {
+    synchronized (this) {
+      // using synchronized to avoid concurrent read that makes readDataSize not correct
+      long before = readDataSize.get();
+      Chunk chunk = super.readMemChunk(metaData);
+      long dataSize = readDataSize.get() - before;
+      CompactionMetrics.getInstance()
+          .recordReadInfo(
+              compactionType,
+              readingAlignedSeries
+                  ? CompactionIoDataType.ALIGNED
+                  : CompactionIoDataType.NOT_ALIGNED,
+              dataSize);
+      return chunk;
+    }
+  }
+
+  @Override
+  public TsFileDeviceIterator getAllDevicesIteratorWithIsAligned() throws IOException {
+    long before = readDataSize.get();
+    TsFileDeviceIterator iterator = super.getAllDevicesIteratorWithIsAligned();
+    long dataSize = readDataSize.get() - before;
+    CompactionMetrics.getInstance()
+        .recordReadInfo(compactionType, CompactionIoDataType.METADATA, dataSize);
+    return iterator;
+  }
```

Comment:
When recording the amount of metadata read here, there will be less records. The process of compaction reads the device iteratively. At the beginning, it traverses and reads the entire index tree to obtain the offset of the leaf device node, and then reads all the device information in the node according to the offset when compaction.